Promote files to blocks

gpetiot opened the pull request:
Fix issue #30 
Work in progress: it would be better to replace the "lines" attributes with "parts" (more robust).

gpetiot commented:
ocaml_topexpect should be migrated to 4.06.1 to be used in mdx.

samoht commented:
> Edit: @samoht does this have to be tackled before merging this PR ?

Not that's not necessary. Let me know when the PR is ready to be merged!

samoht commented on bin/promote/main.ml:
```diff
@@ -0,0 +1,139 @@
+(*
+ * Copyright (c) 2018 Thomas Gazagnaire <thomas@gazagnaire.org>
```
You should add your name here :-)

samoht commented on bin/promote/ocaml_topexpect.ml:
```diff
@@ -0,0 +1,513 @@
+module Chunk = struct
```
can you add some headers?

samoht commented on test/promote_file.md:
````diff
@@ -0,0 +1,32 @@
+Mdx can also understand ocaml code blocks:
+
+```ocaml file=toto.ml,part=toto
````
that should be `file=promote.ml` right?

Also you need to add the relevant runes in the `dune` file.

samoht commented on test/promote_file.md.corrected:
```diff
@@ -0,0 +1,64 @@
+Mdx can also understand ocaml code blocks:
```
Other tests use the file extension `.expected` (e.g. that's expected that the generated `.corrected` file is similar to that file).

gpetiot commented:
I rebased the branch against master but it there is still some issues on the promote_file test, I think I missed something with the jbuider to dune migration (weird, CI says it is ok, but not on my own machine).

For now, based on the last modification date of the .md file and the source file (mentioned in a block), for each block:
- if the .md file is more recent, the source file is updated (in place)
- if the source file is more recent, the .md file is updated (.md.corrected generated)

Does this fit your workflow?

samoht commented:
After some thoughts, I think the new `promote` command could work well if the semantics of `mdx test` for blocks with a `file=..` header is modified to match the following:

A typical user interaction would look like:

    $ cat README.md
    ```ocaml file=dir/promote.ml,part=toto
    let x = 34
    let f = 42.3
    let s = "toto"
    let f x u = u x
    ```

### Normal interaction

    $ mdx test
    [ the block is skipped as promote.ml should be tested/compiled separately ]

### Promotion from ml to md

    $ mdx test README.md
    [ ok ]
    [ edit dir/promote.ml ]
    $ mdx test
    [ generate README.ml.corrected ]
    $ dune promote                     # if dune rules are set
    $ mdx promote [--direction=to-ml]  # optional as dune could handle it today

The `--direction` option has a default value of `infer-timestamp` to use timestamp detection to infer the direction of the promotion. We could later add a `infer-vcs` to check which file has changed using `git-diff` if needed.
 
### Promotion from md to ml

    $ mdx test README.md
    [ ok ]
    [ edit README.md ]
    $ mdx test
    [ generate a dir/promote.ml.corrected ]
    $ mdx promote [--direction=to-ml]  # dune promote doesn't work here (I hink) 
                                       # as it needs to know about dir/promote.ml

### Integration with dune

In the longer term, maybe the right way to integrate with dune is to have a special command generating the right test stanza for dune, something like:

```
$ mdx rule README.md
(alias
  (name runtest)
  (deps (x: README.md) (y: dir/promote.ml))
   (action (progn
      (run mdx test %{x})
       (diff %{x} %{x}.corrected)
       (diff %{y} %{y}.corrected)))
```

Note: I have no idea is that rule is enough :-) In any case, we don't need it right now, we check if I can the new commands are enough for RWO already.

samoht commented on bin/cli.ml:
```diff
@@ -40,6 +40,24 @@ let root =
+    ; "to-ml", `To_ml ]
+  in
+  let names = ["direction"] in
+  let docv = Astring.String.concat ~sep:"|" (List.map fst opt_names) in
```
You can just use `String.concat "|"` here (no need to bring astring in)

gpetiot commented on jbuilder/jbuild:
```diff
@@ -0,0 +1,24 @@
+
```
Yes but it seems this is used instead of dune in the RWO v2 repo

samoht replied:
this needs to be updated -- I'll do it at one point

samoht replied:
I am updating RWO to generate a dune file, so this can be updated. Also please move this in `test` and a a test for it.

gpetiot replied:
This file itself is not bringing any value compared to the dune file, so I will delete it. I don't think it is necessary to add a test for the promote mechanism, we just use `dune promote`, I suppose it is already tested in dune, and we already test that the generated .corrected files match the .expected files.

gpetiot commented on Makefile:
```diff
@@ -3,6 +3,9 @@
 	dune build
 
+install:
+	dune install
```
I am just used to do `make install` but we can skip it if nobody else uses it.

samoht replied:
yes please remove this - this is not usually a good idea to `make install` on behalf on opam.

samoht commented:
Is this ready to merge/review?

gpetiot commented:
Yes it is ready

samoht commented on bin/cli.ml:
```diff
@@ -40,6 +40,24 @@ let root =
+  let names = ["direction"] in
+  let docv = String.concat "|" (List.map fst opt_names) in
+  let docv = "{" ^ docv ^ "}" in
+  Arg.(value & opt (some (enum opt_names)) None & info names ~doc ~docv)
```
Set the default to a sensible value here (for instance `Infer_timestamp`) - that will update the man page automatically.

samoht commented on bin/test/main.ml:
```diff
@@ -130,9 +137,65 @@ let run_toplevel_tests c ppf tests t =
+let update_block_with_file ppf t file part =
+  Block.pp_header ppf t;
+  let lines = Mdx_top.lines_of_part ~file ~part in
+  let contents = Astring.String.concat ~sep:"\n" lines in
```
You don't need `Astring` for this.

samoht commented on bin/test/main.ml:
```diff
@@ -130,9 +137,65 @@ let run_toplevel_tests c ppf tests t =
+        Mdx_top.replace_lines_of_part
+          ~file:input_file ~part ~lines in
+      let lines = List.map (Astring.String.concat ~sep:"\n") lines in
+      let lines = Astring.String.concat ~sep:"\n" lines in
```
Same here, you don't need `Astring` for this.

samoht commented on lib/top/mdx_top.mli:
```diff
@@ -26,3 +26,15 @@ val init: verbose:bool -> silent:bool -> verbose_findlib:bool -> unit -> t
     mulitple lines) with the configuration value [t]. *)
+
+val lines_of_part: file:string -> part:string option -> string list
+(** [lines_of_part f p] returns each line of the part [p] in file [f],
```
wrong docstring

samoht commented on lib/top/mdx_top.mli:
```diff
@@ -26,3 +26,15 @@ val init: verbose:bool -> silent:bool -> verbose_findlib:bool -> unit -> t
+  part:string ->
+  lines:string list ->
+  string list list
+(** [replace_lines_of_part f p l] returns the lines of the file [f] where the
```
wrong docstring.

samoht commented on lib/top/ocaml_topexpect.ml:
```diff
@@ -0,0 +1,319 @@
+(*
```
rename this file to something else, as it hasn't anything to do with ocaml_topexexpt anymore. Maybe `part.ml` with 2 exposed functions: `find` and `replace`?

samoht commented:
I am fine to merge this, we can tune the behaviour that we really want after some use.

samoht commented:
Thanks!
